File: problem_0035.py
#!/usr/bin/env python3
"""
The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.
There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.
How many circular primes are there below one million?
"""

from common import prime_generator
from itertools import count
import logging
import sys
import time

logging.basicConfig(
    level=logging.INFO,
    stream=sys.stderr,
    format='%(levelname)s: %(message)s')

def main():
    prime_gen  = prime_generator()
    prime_list = [next(prime_gen)]
    circular_count = 0
    logging.info("Starting")
    for i in count():
        while i >= len(prime_list):
            logging.debug("Extending primes array")
            new_prime = next(prime_gen)
            prime_list.append(new_prime)
        circular = True
        if prime_list[i] >= 1000000: break
        logging.debug("Testing " + str(prime_list[i]))
        for circ in circ_gen(prime_list[i]):
            while circ > prime_list[-1]:
                prime_list.append(next(prime_gen))
            if circ not in prime_list:
                circular = False
                break
        if circular:
            logging.info("This prime is circular " + str(prime_list[i]))
            circular_count += 1
    print('There are {} circular primes'.format(circular_count))

def circ_gen(number):
    str_num = str(number)
    for i in range(1,len(str_num)):
        yield int(str_num[i:] + str_num[:i])

if __name__ =="__main__":
    start = time.time()
    main()
    print('That took {:4.2f} seconds'.format(time.time() - start))

File: problem_0035.hs
{-# LANGUAGE DeriveDataTypeable, RecordWildCards #-}

-- The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.
-- There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.
-- How many circular primes are there below one million?

import Test.HUnit ((@=?), assertBool, runTestTT, Test(..))
import Text.Printf (printf)
import System.Console.CmdArgs
import Data.Time (getCurrentTime, diffUTCTime)

data EulerArgs = 
    AdHoc { limit::Integer }
    | Euler 
    | UnitTest
    deriving (Show, Data, Typeable)

problem0035 :: Integer -> [Integer]
problem0035 limit = filter isCircularPrime $ takeWhile (<limit) primes

isCircularPrime :: Integer -> Bool
isCircularPrime num = all (isPrime . read) rotations
    where
        strNum = show num 
        rotations = map (rotate strNum) [0..(length strNum)]

isCircularPrimeTest = [
    True @=? isCircularPrime 197]

rotate :: [a] -> Int -> [a]
rotate xs n = zipWith const (drop (if n < 0 then n' else n) $ cycle xs) xs
    where
        n' = mod n $ length xs

rotateTest = [
    "3412" @=? rotate "1234" (-10) ,
    "3412" @=? rotate "1234" (-2) ,
    "3412" @=? rotate "1234" 6 ,
    "1234" @=? rotate "1234" 4 ,
    "2341" @=? rotate "1234" 1 ,
    "1234" @=? rotate "1234" 0]

primes :: [Integer]
primes = 2 : [n | n <- [3,5..], isPrime n]

primesTest = [
    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29] @=? (take 10 primes),
    [2, 3] @=? (take 2 primes),
    2 @=? (head primes)]

isPrime :: Integer -> Bool
isPrime 0 = False
isPrime n = all noRemainer primes'
    where
        primes' = takeWhile (\p -> p * p <= n) primes
        noRemainer d = 0 /= rem n d

isPrimeTest = [
    assertBool "9 is not prime" (not $ isPrime 9),
    assertBool "6 is not prime" (not $ isPrime 6),
    assertBool "4 is not prime" (not $ isPrime 4),
    assertBool "0 is not prime" (not $ isPrime 0),
    assertBool "3 is prime" (isPrime 3),
    assertBool "2 is prime" (isPrime 2)]    

unitTests = map TestCase $
    isPrimeTest ++
    primesTest ++
    rotateTest ++
    isCircularPrimeTest

exec :: EulerArgs -> IO ()
exec AdHoc{..}= do
    let answer = problem0035 limit
    printf "Below %d there are %d circular primes:" limit (length answer)
    mapM_ print answer
exec Euler = do
    let answer = length $ problem0035 1000000
    printf "Answer: %d\n" answer
exec UnitTest = do 
    runTestTT $ TestList unitTests
    return ()

main :: IO ()
main = do
    args <- cmdArgs $ modes [AdHoc{ limit = 100 }, Euler, UnitTest]
    start <- getCurrentTime
    exec args
    stop <- getCurrentTime
    print $ diffUTCTime stop start
